File: hs/src/Q800/Q844.hs
module Q800.Q844
(
    backspaceComparePlanA
    )
where

backspaceComparePlanA :: String -> String -> Bool
backspaceComparePlanA s c =
                    let ss = m "" s
                        cs = m "" c
                    in ss == cs
                    where
                        m str [] = str
                        m str (x:xs) = if x == '#' then m (init str) xs
                                       else m (str++[x]) xs

File: golang/q800/q844.go
package q800

func backspaceComparePlanA(s string, t string) bool {
	m := func(str string) string {
		r := make([]rune, 0, len(str))
		for _, v := range str {
			if v == '#' {
				if len(r) > 0 {
					r = r[:len(r)-1]
				}
			} else {
				r = append(r, v)
			}
		}
		return string(r)
	}
	return m(s) == m(t)
}
